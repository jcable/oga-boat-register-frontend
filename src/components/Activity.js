import React, { useState } from 'react';
import Box from '@material-ui/core/Box';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import Button from '@material-ui/core/Button';
import DescriptionIcon from '@material-ui/icons/Description';
import ContactMailIcon from '@material-ui/icons/ContactMail';
import FastForward from '@material-ui/icons/FastForward';
import TableChart from '@material-ui/icons/TableChart';
import BoatIcon from './boaticon';

export default function Activity({ classes, onCancel, onStart }) {
  const [selectedIndex, setSelectedIndex] = useState(1);

  const handleListItemClick = (event, index) => {
    setSelectedIndex(index);
  };

  const handleClickCancel = () => {
    onCancel();
  };

  const handleClickStart = () => {
    onStart(selectedIndex);
  };

  return (
    <>
        <Paper className={classes.paper}>
          <Typography component="h1" variant="h5" align="center">
            What do you want to do today?
          </Typography>
	  <List>
           <ListItem button selected={selectedIndex === 0} onClick={(event) => handleListItemClick(event, 0)} >
            <ListItemIcon><DescriptionIcon /></ListItemIcon>
            <ListItemText primary="Update the short and long descriptions" />
           </ListItem>
           <ListItem button selected={selectedIndex === 1} onClick={(event) => handleListItemClick(event, 1)} >
            <ListItemIcon><BoatIcon /></ListItemIcon>
            <ListItemText primary="Update rig and basic dimensions" />
           </ListItem>
           <ListItem button selected={selectedIndex === 2} onClick={(event) => handleListItemClick(event, 2)} >
            <ListItemIcon><TableChart /></ListItemIcon>
            <ListItemText primary="Create or update a handicap" />
           </ListItem>
           <ListItem button selected={selectedIndex === 3} onClick={(event) => handleListItemClick(event, 3)} >
            <ListItemIcon><ContactMailIcon /></ListItemIcon>
            <ListItemText primary="Update ownership and/or for-sale information" />
           </ListItem>
           <ListItem button selected={selectedIndex === 4} onClick={(event) => handleListItemClick(event, 4)} >
            <ListItemIcon><FastForward /></ListItemIcon>
            <ListItemText primary="Do a full review/update" />
           </ListItem>
	  </List>
          <Box display="flex" justifyContent="flex-end">
            <Box m={1}>
              <Button variant="outlined" color="primary" onClick={handleClickCancel}>
                Cancel
              </Button>
            </Box>
            <Box m={1}>
              <Button variant="outlined" color="primary" onClick={handleClickStart}>
              Let's go
              </Button>
            </Box>
          </Box>
        </Paper>
    </>
  );
}
