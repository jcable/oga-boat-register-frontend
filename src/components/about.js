import React from 'react';
import Typography from '@material-ui/core/Typography'

function About() {

    return (
        <Typography>In the early 1970s John Scarlett, the first OGA Secretary, started collecting data on members' boats. The records have grown to include nearly 3,500 craft, forming a unique record. Mainly gaff rigged, old and new, workboats and yachts, the Register also contains classic bermudan yachts and motorboats. Data have been stored electronically since the 1980s and in 2016 the OGA launched the register online. Whilst primarily a record of members boats the register has become an important historical record of working and leisure craft and are happy to receive details about interesting boats regardless of their association with the OGA. OGA Members can list boats 'For Sale' at no additional charge.


You are browsing the latest new version of the online register. The new register will feature separate online forms for adding a new boat, updating an existing entry and putting a boat up for sale. If you prefer you can still download a form to print, fill in by hand and post to us.


An important feature of a boat's entry are photographs. The online forms allow you to upload digital photographs but if you only have prints, slides or negatives please contact the boat register editors.


If you have bought or sold a boat let us know so we can update the ownership details

Click here to browse the Boat Register

Please contact the Boat Register Editor with feedback.

Boat register form
If your boat is not in the register, we would love to include it. Contact the Boat Register Editor with a few details.

We have a new online form and we would love you to try it out. If you prefer to fill in a paper form, you can download one and post it to us with pictures and we will add it to the database.

If you find either of the forms confusing get in touch and we will help you.

Once we've confirmed your boat isn't already on the register we will create an entry. If you are an OGA member you will be able to fill in the details and add photographs online.

If you have bought or sold a boat let us know so we can update the ownership details.

or search the Boat Register.</Typography>
    );
}

export default About