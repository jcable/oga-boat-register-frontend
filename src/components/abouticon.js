import React from 'react';
import SvgIcon from '@material-ui/core/SvgIcon';

  function AboutIcon(props) {
    return (
      <SvgIcon {...props}>
        <g>
        <path id="bubble" d="M 0 11 L 0 0 L 24 0 L 24 22 L 0 22 Z M 3.601562 21.546875 C 5.101562 21.347656 7.238281 20.6875 9.238281 19.800781 L 9.644531 19.621094 L 9.992188 19.683594 C 10.773438 19.824219 11.191406 19.855469 12.171875 19.855469 C 13.269531 19.855469 13.707031 19.8125 14.667969 19.621094 C 18.410156 18.875 21.472656 16.371094 22.648438 13.105469 C 23.023438 12.070312 23.15625 11.28125 23.15625 10.09375 C 23.15625 9.054688 23.074219 8.441406 22.808594 7.5625 C 21.753906 4.035156 18.570312 1.316406 14.589844 0.542969 C 13.699219 0.371094 13.195312 0.324219 12.171875 0.324219 C 11.148438 0.324219 10.644531 0.371094 9.753906 0.542969 C 6.179688 1.242188 3.183594 3.539062 1.886719 6.582031 C 0.75 9.261719 0.976562 12.324219 2.5 14.789062 C 2.867188 15.390625 3.359375 15.996094 3.960938 16.601562 L 4.546875 17.183594 L 4.433594 17.507812 C 4.113281 18.421875 3.59375 19.0625 2.386719 20.050781 C 1.207031 21.015625 1.03125 21.261719 1.433594 21.417969 C 1.929688 21.605469 2.75 21.652344 3.601562 21.546875 Z M 3.601562 21.546875 "/>
        <path id="bubbledot" d="M 10.859375 15.609375 C 10.300781 15.449219 10 15.21875 9.777344 14.78125 C 9.660156 14.554688 9.65625 14.527344 9.660156 13.894531 C 9.664062 13.34375 9.683594 13.1875 9.777344 12.835938 C 10.554688 9.972656 10.589844 9.828125 10.535156 9.53125 C 10.472656 9.214844 10.300781 8.921875 10.144531 8.863281 C 9.90625 8.78125 9.453125 8.808594 9.132812 8.925781 C 8.972656 8.980469 8.835938 9.019531 8.824219 9.011719 C 8.8125 9.003906 8.839844 8.859375 8.882812 8.691406 L 8.953125 8.394531 L 9.28125 8.265625 C 10.085938 7.945312 10.65625 7.816406 11.226562 7.808594 C 11.710938 7.808594 12.007812 7.871094 12.375 8.046875 C 12.882812 8.292969 13.09375 8.613281 13.167969 9.25 C 13.246094 9.875 13.175781 10.351562 12.75 11.921875 C 12.230469 13.839844 12.183594 14.148438 12.371094 14.421875 C 12.542969 14.691406 13.265625 14.761719 13.753906 14.554688 C 14.058594 14.425781 14.050781 14.40625 13.878906 14.847656 L 13.78125 15.105469 L 13.222656 15.3125 C 12.207031 15.691406 11.457031 15.78125 10.859375 15.609375 Z M 10.859375 15.609375 "/>
        <path id="bubblei" d="M 12.3125 6.398438 C 11.578125 6.242188 11.097656 5.707031 11.089844 5.027344 C 11.082031 4.558594 11.226562 4.222656 11.542969 3.941406 C 12.039062 3.503906 12.691406 3.425781 13.304688 3.730469 C 13.839844 3.996094 14.109375 4.414062 14.109375 4.988281 C 14.109375 5.320312 14.03125 5.5625 13.824219 5.839844 C 13.703125 6.007812 13.601562 6.085938 13.320312 6.222656 C 12.9375 6.410156 12.628906 6.460938 12.3125 6.398438 Z M 12.3125 6.398438 "/>
        </g>
      </SvgIcon>
    );
  }

export default AboutIcon