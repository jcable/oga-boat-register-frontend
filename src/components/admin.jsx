import React from 'react';
import Typography from '@material-ui/core/Typography';

export default function Admin({ classes, boat }) {
    return (<Typography>editors get special treatment</Typography>);
}