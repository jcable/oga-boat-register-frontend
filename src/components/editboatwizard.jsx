import React, { useEffect, useState } from "react";
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import enGB from "date-fns/locale/en-GB";
import { v4 as uuidv4 } from 'uuid';
import { create, formatters } from 'jsondiffpatch';
import FormRenderer from "@data-driven-forms/react-form-renderer/form-renderer";
import componentMapper from "@data-driven-forms/mui-component-mapper/component-mapper";
import FormTemplate from "@data-driven-forms/mui-component-mapper/form-template";
import Box from "@mui/material/Box";
import Dialog from "@mui/material/Dialog";
import CircularProgress from '@mui/material/CircularProgress';
import {
  designerItems,
  builderItems,
  constructionItems,
  designClassItems,
} from "./ddf/util";
import { steps as handicap_steps } from "./Handicap";
import {
  yearItems,
  homeItems,
  registrationForm,
  referencesItems,
  salesSteps,
  ownerShipsFields,
  sellingDataFields,
  doneFields,
  hullFields,
  descriptionsItems,
  basicFields,
} from "./ddf/SubForms";
import Typography from "@mui/material/Typography";
import { getPicklists } from './boatregisterposts';
import HtmlEditor from './tinymce';
import { boatm2f, boatf2m, boatDefined } from "../util/format";

const defaultSchema = (pickers) => {
  return {
    fields: [
      {
        title: "Update Boat Record",
        component: 'wizard',
        name: "boat",
        fields: [
          {
            name: "basic-step",
            nextStep: "descriptions-step",
            fields: [
              {
                component: 'text-field',
                name: "ddf.owner",
                label: "is current owner",
                hideField: true,
              },
              {
                component: 'text-field',
                name: "ddf.editor",
                label: "is editor",
                hideField: true,
              },
              {
                component: 'sub-form',
                name: "basic.form",
                title: "Basic Details",
                fields: basicFields(pickers),
              },
            ],
          },
          {
            name: "descriptions-step",
            nextStep: 'build-step',
            fields: descriptionsItems,
          },
          {
            name: "build-step",
            nextStep: "design-step",
            fields: [
              {
                title: "Build",
                name: "build",
                component: 'sub-form',
                fields: [
                  ...yearItems,
                  {
                    component: 'text-field',
                    name: "place_built",
                    label: "Place Built",
                  },
                  ...builderItems(pickers),
                  {
                    component: 'text-field',
                    name: "hin",
                    label: "Hull Identification Number (HIN)",
                  },
                ],
              },
            ],
          },
          {
            name: "design-step",
            nextStep: "basic-dimensions-step",
            fields: [
              {
                title: "Design",
                name: "design",
                component: 'sub-form',
                fields: [
                  ...designerItems(pickers),
                  ...designClassItems(pickers),
                ],
              },
            ],
          },
          {
            name: "basic-dimensions-step",
            nextStep: "references-step",
            fields: [
              {
                title: "Basic Dimensions",
                name: "basic-dimensions",
                component: 'sub-form',
                fields: [
                  {
                    component: 'text-field',
                    name: "handicap_data.length_on_deck",
                    label: "Length on deck (LOD) (decimal feet)",
                    type: "number",
                    dataType: 'float',
                  },
                  {
                    component: 'text-field',
                    name: "handicap_data.length_on_waterline",
                    label: "Waterline Length {LWL) (decimal feet)",
                    type: "number",
                    dataType: 'float',
                  },
                  {
                    component: 'text-field',
                    name: "handicap_data.beam",
                    label: "Beam (decimal feet)",
                    type: "number",
                    dataType: 'float',
                    isRequired: true,
                    validate: [
                      {
                        type: 'required',
                      },
                      /*{
                        type: 'min-number-value',
                        threshold: 1,
                      }*/
                    ],
                  },
                  {
                    component: 'text-field',
                    name: "handicap_data.draft",
                    label: "Minumum Draft (decimal feet)",
                    type: "number",
                    dataType: 'float',
                    isRequired: true,
                    validate: [
                      {
                        type: 'required',
                      },
                      /*{
                       type: 'min-number-value',
                       threshold: 1
                      }*/
                    ],
                  },
                  {
                    component: 'text-field',
                    name: "air_draft",
                    label: "Air Draft (decimal feet)",
                    type: "number",
                    dataType: 'float',
                    validate: [
                      /*{
                        type: 'min-number-value',
                        threshold: 1
                      }*/
                    ],
                  },
                  {
                    component: 'plain-text',
                    name: "ddf.h1",
                    label: "LOD, LWL, beam and draft affect handicaps",
                  },
                ],
              },
            ],
          },
          {
            name: "references-step",
            nextStep: "previousnames-step",
            fields: [
              {
                name: "references",
                title: "References",
                component: 'sub-form',
                fields: referencesItems,
              },
            ],
          },
          {
            name: "previousnames-step",
            nextStep: "locations-step",
            fields: [
              {
                label: "New name",
                component: 'text-field',
                name: "ddf.new_name",
                description: 'if you have changed the name, enter the new name here.'
              },
              {
                label: "Previous names",
                component: 'field-array',
                name: "previous_names",
                fields: [{ component: "text-field" }],
              },
            ],
          },
          {
            name: "locations-step",
            nextStep: "registrations-step",
            fields: [
              {
                title: "Locations",
                name: "locations",
                component: 'sub-form',
                fields: homeItems,
              },
            ],
          },
          {
            name: "registrations-step",
            nextStep: "construction-step",
            fields: [registrationForm],
          },
          {
            name: "construction-step",
            nextStep: 'hull-step',
            fields: [
              {
                name: "construction",
                title: "Construction",
                component: 'sub-form',
                fields: constructionItems(pickers),
              },
            ],
          },
          {
            name: "hull-step",
            nextStep: 'skip-handicap-step',
            fields: hullFields,
          },
          {
            name: "skip-handicap-step",
            nextStep: {
              when: "ddf.skip-handicap",
              stepMapper: {
                1: "handicap-step",
                2: "own-step",
              },
            },
            fields: [
              {
                name: "skip-handicap",
                title: "Handicaps",
                component: 'sub-form',
                fields: [
                  {
                    component: 'radio',
                    name: "ddf.skip-handicap",
                    initialValue: "1",
                    label: 'Do you want to create or update a handicap?',
                    helperText: "There are some mandatory fields in the handicap section. If you don't need a handicap right now, you can skip this part.",
                    validate: [{ type: 'required' }],
                    options: [
                      { label: "I want to add or check handicap data", value: "1" },
                      { label: "I'll leave it for now", value: "2" }
                    ],
                  },
                ],
              },
            ],
          },
          ...handicap_steps('handicap-step', 'own-step'),
          {
            name: 'own-step',
            nextStep: ({ values }) => {
              if (values.ddf.owner || values.ddf.editor) {
                if (values.ddf.update_sale === 'update') {
                  return 'update-sell-step';
                }
                return 'query-sell-step';
              }
              return 'done-step';
            },
            fields: ownerShipsFields,
          },
          {
            name: 'query-sell-step',
            nextStep: {
              when: "ddf.confirm_for_sale",
              stepMapper: {
                true: 'sell-step',
                false: 'done-step',
              },
            },
            fields: [
              {
                component: 'checkbox',
                label: 'I want to sell this boat',
                name: 'ddf.confirm_for_sale',
                helperText: 'check if you want to put this boat up for sale',
                resolveProps: (props, { meta, input }, formOptions) => {
                  const { values } = formOptions.getState();
                  return {
                    initialValue: !!values.ddf.current_sales_record,
                    isReadOnly: !(values.ddf.owner || values.ddf.editor)
                  }
                },
              },
            ],
          },
          {
            name: 'sell-step',
            nextStep: 'done-step',
            fields: sellingDataFields,
          },
          ...salesSteps('update-sell-step', 'done-step'),
          {
            name: "done-step",
            fields: doneFields,
          },
        ],
      },
    ],
  };
};

export function boatdiff(before, after) {
  const cj = create({
    objectHash: (obj) => `${obj.id || obj.name}-${obj.start}`,
    textDiff: { minLength: 60000 }, // prevent textdiff not supported by the RFC formatter
  });
  return cj.diff(before, after);
}

export function prepareInitialValues(boat, user) {
  const ownerids = boat.ownerships?.filter((o) => o.current)?.map((o) => o.id) || [];
  const goldId = user?.['https://oga.org.uk/id'];
  const editor = (user?.['https://oga.org.uk/roles'] || []).includes('editor');
  const owner = ownerids.includes(goldId);
  const { name, oga_no, id, image_key, for_sales, for_sale_state, ...rest } = boat;

  const sortedsales = [...(for_sales || [])];
  // descending
  sortedsales.sort((a, b) => Date.parse(b.created_at) - Date.parse(a.created_at));

  const ddf = { name, oga_no, id, image_key, owner, editor };

  if (boat.selling_status === 'for_sale') {
    ddf.current_sales_record = sortedsales[0];
    ddf.update_sale = 'update';
  } else {
    ddf.current_sales_record = {
      created_at: new Date().toISOString(),
      asking_price: 0, 
      sales_text: '', 
      flexibility: 'normal',
    };
    ddf.update_sale = 'unsell';
  }

  const initialValues = {
    email: user?.email || '',
    ...boatm2f(rest),
    ddf,
  };

  console.log('DDF', ddf);

  ['builder', 'designer'].forEach((key) => {
    const val = initialValues[key];
    if (Array.isArray(val)) {
      initialValues[key] = val.map((v) => v.name);
    } else if (val) {
      initialValues[key] = [val.name]
    } else {
      initialValues[key] = [];
    }
  });

  ['design_class'].forEach((key) => {
    initialValues[key] = initialValues[key]?.name;
  });

  return initialValues;

}

export function salesChanges(ddf, original_for_sales = []) {
  console.log(salesChanges, ddf, original_for_sales);
  let selling_status = 'not_for_sale';

  switch (ddf.update_sale) {
    case 'unsell':
    case 'sold':
      selling_status = 'not_for_sale';
      break;
    case 'update':
      selling_status = 'for_sale';
      break;
    default:
      if (ddf.confirm_for_sale && ddf.current_sales_record) {
        selling_status = 'for_sale';
      } else {
        console.log('handleSubmit unexpected update_sale value', ddf);
      }
  }
  
  const other_sales = original_for_sales.filter((fs) => fs.created_at !== ddf.current_sales_record.created_at);

  const for_sales = [ddf.current_sales_record, ...other_sales];

  // descending
  for_sales.sort((a, b) => Date.parse(b.created_at) - Date.parse(a.created_at));

  return { for_sales, selling_status };
}

export function prepareModifiedValues(values, { name, oga_no, id, image_key, for_sales }, pickers) {
  const { ddf, email, ...submitted } = values;

  const newItems = Object.fromEntries(
    ['builder', 'designer', 'design_class']
      .filter((key) => ddf[`new_${key}`])
      .map((key) => {
        return { name: ddf[`new_${key}`], id: uuidv4() };
      })
  );

  function name2object(value, picker, newItem) {
    if (value?.name) {
      return value;
    }
    if (value === newItem?.name) {
      return newItem;
    }
    const r = picker.find((p) => p.name === value);
    if (r) {
      return r;
    }
    console.log(`${value} is missing in picklist, making a new uuid`);
    return { name: value, id: uuidv4() };
  }

  const design_class = name2object(values.design_class, pickers['design_class'], newItems['design_class']);

  const builder = values.builder.map((v) =>  name2object(v, pickers['builder'], newItems['builder']));
  const designer = values.designer.map((v) =>  name2object(v, pickers['designer'], newItems['designer']));

  const boat = {
    ...boatf2m(submitted),
    name: ddf.new_name || name,
    previous_names: [
      ...((ddf.new_name && [name]) || []),
      ...(submitted.previous_names || []),
    ],
    oga_no, id, image_key,
    ...salesChanges(ddf, for_sales),
    builder,
    designer,
    design_class,
  };

  return { boat: boatDefined(boat), newItems, email };
}

export default function EditBoatWizard({ boat, user, open, onCancel, onSubmit, schema }) {

  const [pickers, setPickers] = useState();

  const handleSubmit = (values, formApi) => {

    const { newItems, email, boat: modifiedBoat } = prepareModifiedValues(values, boat, pickers);

    const fulldelta = formatters.jsonpatch.format(boatdiff(boat, modifiedBoat));

    onSubmit(
      fulldelta,
      newItems,
      modifiedBoat,
      email,
    );
  }

  useEffect(() => {
    if (!pickers) {
      getPicklists().then((r) => {
        setPickers(r)
      }).catch((e) => console.log(e));
    }
  }, [pickers]);

  if (!pickers) return <CircularProgress />;

  if (!open) return '';

  const activeSchema = schema || defaultSchema(pickers);

  return (
    <Dialog
      open={open}
      aria-labelledby="form-dialog-title"
    >
      <Box sx={{ marginLeft: '1.5rem', marginTop: '1rem' }}>
        <Typography variant="h5" >Update {boat.name} ({boat.oga_no})</Typography>
      </Box>
      <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={enGB}>
        <FormRenderer
          componentMapper={{
            ...componentMapper,
            html: HtmlEditor,
          }}
          FormTemplate={(props) => (
            <FormTemplate {...props} showFormControls={false} />
          )}
          schema={activeSchema}
          onSubmit={handleSubmit}
          onCancel={onCancel}
          initialValues={prepareInitialValues(boat, user)}
          subscription={{ values: true }}
        />
      </LocalizationProvider>

    </Dialog>
  );
}
