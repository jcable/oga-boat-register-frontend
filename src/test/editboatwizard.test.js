import fs from 'fs';
import { buildChanges, includes, elements_matching_by_field, nameChanges, recursiveUpdate, salesChanges, updateOwnerships, updateOwnership } from '../components/editboatwizard';
import { channel } from 'diagnostics_channel';

const robinetta = JSON.parse(fs.readFileSync('src/test/robinetta.json'));
const pickers = {
  design_class: [{ "id": "c05a2174-4b46-410f-829f-894193f020c3", "name": "Falmouth Bass Boat" }],
  designer: [],
  builder: [],
};

test('nameChanges empty none', () => {
  expect(nameChanges({}, undefined)).toBe(undefined);
});

test('nameChanges non empty none', () => {
  expect(nameChanges({ name: 'X' }, undefined)).toBe(undefined);
});

test('nameChanges no previous change', () => {
  expect(nameChanges({ name: 'Spray' }, 'Cloud')).toMatchSnapshot();
});

test('nameChanges with previous change', () => {
  expect(nameChanges({
    name: 'Spray',
    previous_names: ['Victoria'],
  }, 'Cloud')).toMatchSnapshot();
});

test('salesChanges none', () => {
  expect(salesChanges({}, {}, {}, {})).toMatchSnapshot();
});

test('salesChanges no change', () => {
  expect(salesChanges({}, {}, {}, {})).toMatchSnapshot();
});

test('buildChanges none', () => {
  const { removed, ...changes } = buildChanges({}, {}, {}, {}, {}, {});
  expect(changes).toMatchSnapshot();
});

test('buildChanges no change', () => {
  const { removed, ...changes1 } = buildChanges({}, robinetta, {}, {}, {}, {});
  expect(changes1).toStrictEqual({});
  let changes2;
  {
    const { removed, ...changes } = buildChanges({}, {
      design_class: {
        id: 'c05a2174-4b46-410f-829f-894193f020c3',
        name: 'Falmouth Bass Boat',
      },
    }, {}, {}, {}, pickers);
    changes2 = changes;
  }
  expect(changes2).toStrictEqual({});
});

test('buildChanges name change', () => {
  const { removed, ...changes } = buildChanges({},
    {
      name: 'Spray',
      previous_names: ['Victoria'],
    },
    {},
    { new_name: 'Cloud' },
    {},
    {});
  expect(changes).toMatchSnapshot();
});

test('recursiveUpdate no change', () => {
  expect(recursiveUpdate(robinetta, {})).toStrictEqual(robinetta);
});

test('recursiveUpdate add air draft', () => {
  expect(recursiveUpdate(robinetta, {
    air_draft: 4.0
  })).toStrictEqual({ ...robinetta, air_draft: 4.0 });
});

test('includes', () => {
  expect(includes([], true)).toBe(false);
  expect(includes([1], 0)).toBe(false);
  expect(includes([1], 1)).toBe(true);
  expect(includes([{ x: 1 }], { x: 1 })).toBe(true);
});

test('match by field', () => {
  const o = [
    { start: 1986, end: 2007, name: 'Mike Garnham', share: 64 },
    { id: 559, member: 5004, share: 32, start: 2007, end: 2025 },
    { id: 1219, member: 5004, share: 32, start: 2007, end: 2025 },
  ];
  const n = [{ name: 'John Smith', start: 2025, share: 64, current: true }];
  expect(elements_matching_by_field(o, n, 'id')).toStrictEqual([]);
  expect(elements_matching_by_field(o, [o[0]], 'id')).toStrictEqual([]);
  expect(elements_matching_by_field(o, [o[0]], 'name')).toStrictEqual([o[0]]);
  expect(elements_matching_by_field(o, [o[0]], 'id')).toStrictEqual([]);
  expect(elements_matching_by_field(o, [o[1]], 'id')).toStrictEqual([o[1]]);
  const r = elements_matching_by_field(o, [o[1]], 'member');
  expect(r).toStrictEqual([o[1], o[2]]);
});

test('update-ownerships 0', () => {
  expect(updateOwnerships([], [])).toStrictEqual([]);
  const o = [
    { id: 559, member: 5004, share: 32, start: 2007, end: 2025 },
    { id: 1219, member: 5004, share: 32, start: 2007, end: 2025 },
  ];
  expect(updateOwnerships(o, [])).toStrictEqual(o);
});

test('update-ownerships 1', () => {
  const o = [
    { start: 1986, end: 2007, name: 'Mike Garnham', share: 64 },
    { id: 559, member: 5004, share: 32, start: 2007, end: 2025 },
    { id: 1219, member: 5004, share: 32, start: 2007, end: 2025 },
  ];
  const n = [{ name: 'John Smith', start: 2025, share: 64, current: true }];
  expect(updateOwnerships(o, n)).toStrictEqual([...o, ...n]);
});

test('update_ownership', () => {
  const o = { id: 559, member: 5004, share: 32, start: 2007, current: true };
  const n = { id: 559, member: 5004, share: 32, start: 2007, end: 2025 };
  expect(updateOwnership(o, n)).toStrictEqual(n);
  const n2 = { id: 559, member: 5004, start: 2007, end: 2025 };
  expect(updateOwnership(o, n2)).toStrictEqual(n);
});

test('update-ownerships 2', () => {
  const o = [
    { start: 1986, end: 2007, name: 'Mike Garnham', share: 64 },
    { id: 559, member: 5004, share: 32, start: 2007 },
    { id: 1219, member: 5004, share: 32, start: 2007 },
  ];
  const n = [
    { id: 559, member: 5004, share: 32, start: 2007, end: 2025 },
    { id: 1219, member: 5004, share: 32, start: 2007, end: 2025 },
    { name: 'John Smith', start: 2025, share: 64, current: true },
  ];
  expect(updateOwnerships(o, n)).toStrictEqual([
    { start: 1986, end: 2007, name: 'Mike Garnham', share: 64 }, ...n]);
});

test('recursiveUpdate new owner', () => {
  const n = [{ name: 'John Smith', start: 2025, share: 64, current: true }];
  const changes = { ownerships: n };
  // console.log(JSON.stringify(robinetta, null, 2));
  const updated = recursiveUpdate(robinetta, changes);
  // console.log(JSON.stringify(updated, null, 2));
  const q = { ...robinetta };
  delete q.ownerships;
  const u = { ...updated };
  delete u.ownerships;
  expect(q).toEqual(u);
  expect(updated.ownerships).toStrictEqual([...robinetta.ownerships, ...n]);
});

test('recursiveUpdate new owner', () => {
  const n = { name: 'John Smith', start: 2025, share: 64, current: true };
  const changes = {
    ownerships: [
      { id: 559, member: 5004, share: 32, start: 2007, end: 2025 },
      { id: 1219, member: 5004, share: 32, start: 2007, end: 2025 },
      n,
    ]
  };
  // console.log(JSON.stringify(robinetta, null, 2));
  const updated = recursiveUpdate(robinetta, changes);
  // console.log(JSON.stringify(updated, null, 2));
  const q = { ...robinetta };
  delete q.ownerships;
  const u = { ...updated };
  delete u.ownerships;
  expect(q).toEqual(u);
  expect(updated.ownerships).toMatchSnapshot();
});

test('handle submit 1', () => {
  const boat = { "air_draft": 3.048, "construction_details": "Built-in bouyancy", "construction_material": "grp", "construction_method": "cold_moulded", "design_class": { "id": "c05a2174-4b46-410f-829f-894193f020c3", "name": "Falmouth Bass Boat" }, "draft": 1.219, "for_sale_state": { "text": "not_for_sale" }, "for_sales": [], "generic_type": "Dinghy", "handicap_data": { "beam": 1.829 }, "home_country": "UK", "home_port": "Bristol", "hull_form": "centre-board dinghy", "id": "5598c94d-4720-4423-a575-44b0845a308b", "image_key": "ddgWxx", "length_on_deck": 4.877, "mainsail_type": "gunter", "name": "Test Boat Record Entry", "oga_no": 9001, "ownerships": [{ "current": true, "id": 35000, "member": 6583, "share": 64, "name": "John B. Sloop" }], "place_built": "Falmouth", "rig_type": "Yawl", "short_description": "<p>Open sailing dinghy.</p>", "spar_material": "wood", "year": 1985, "year_is_approximate": true };
  const ddf = { "can_sell": true, "skip-handicap": "2", "confirm_for_sale": false };
  const submitted = { "mainsail_type": "gunter", "rig_type": "Yawl", "generic_type": "Dinghy", "short_description": "<p>Open sailing dinghy.</p>", "year": 1985, "year_is_approximate": true, "place_built": "Falmouth", "design_class": { "id": "c05a2174-4b46-410f-829f-894193f020c3", "name": "Falmouth Bass Boat" }, "handicap_data": { "length_on_deck": 14, "beam": 6.001, "draft": 1 }, "air_draft": 10, "home_country": "UK", "home_port": "Bristol", "construction_material": "grp", "spar_material": "wood", "construction_details": "Built-in bouyancy", "hull_form": "centre-board dinghy", "ownerships": [{ "current": true, "id": 35000, "member": 6583, "share": 64, "name": "John B. Sloop", "start": 2021 }] };
  const changes = { "handicap_data": { "length_on_deck": 4.267, "draft": 0.305 }, "ownerships": [{ "current": true, "id": 35000, "member": 6583, "share": 64, "name": "John B. Sloop", "start": 2021 }] };
  const newItems = {};
  // expect(buildChanges({}, boat, submitted, ddf, newItems, pickers)).toStrictEqual(changes);
  expect(recursiveUpdate(boat, changes, ['construction_method'])).toMatchSnapshot();


});