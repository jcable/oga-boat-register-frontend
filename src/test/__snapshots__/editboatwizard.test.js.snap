// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildChanges name change 1`] = `
Object {
  "name": "Cloud",
  "previous_names": Array [
    "Spray",
    "Victoria",
  ],
}
`;

exports[`buildChanges none 1`] = `Object {}`;

exports[`handle submit 1 1`] = `
Object {
  "air_draft": 3.048,
  "construction_details": "Built-in bouyancy",
  "construction_material": "grp",
  "design_class": Object {
    "id": "c05a2174-4b46-410f-829f-894193f020c3",
    "name": "Falmouth Bass Boat",
  },
  "draft": 1.219,
  "for_sale_state": Object {
    "text": "not_for_sale",
  },
  "for_sales": Array [],
  "generic_type": "Dinghy",
  "handicap_data": Object {
    "beam": 1.829,
    "draft": 0.305,
    "length_on_deck": 4.267,
  },
  "home_country": "UK",
  "home_port": "Bristol",
  "hull_form": "centre-board dinghy",
  "id": "5598c94d-4720-4423-a575-44b0845a308b",
  "image_key": "ddgWxx",
  "length_on_deck": 4.877,
  "mainsail_type": "gunter",
  "name": "Test Boat Record Entry",
  "oga_no": 9001,
  "ownerships": Array [
    Object {
      "current": true,
      "id": 35000,
      "member": 6583,
      "share": 64,
      "start": 2021,
    },
  ],
  "place_built": "Falmouth",
  "rig_type": "Yawl",
  "short_description": "<p>Open sailing dinghy.</p>",
  "spar_material": "wood",
  "year": 1985,
  "year_is_approximate": true,
}
`;

exports[`nameChanges no previous change 1`] = `
Object {
  "name": "Cloud",
  "previous_names": Array [
    "Spray",
  ],
}
`;

exports[`nameChanges with previous change 1`] = `
Object {
  "name": "Cloud",
  "previous_names": Array [
    "Spray",
    "Victoria",
  ],
}
`;

exports[`recursiveUpdate new owner 1`] = `
Array [
  Object {
    "end": 1946,
    "name": "Denys Rayner",
    "share": 64,
    "start": 1937,
  },
  Object {
    "end": 1949,
    "name": "Audrey Parker",
    "share": 32,
    "start": 1946,
  },
  Object {
    "end": 1949,
    "name": "Henry Parker",
    "share": 32,
    "start": 1946,
  },
  Object {
    "end": 1985,
    "name": "Nigel Heriot",
    "share": 64,
    "start": 1949,
  },
  Object {
    "end": 1986,
    "name": "Frank Driscoll",
    "share": 64,
    "start": 1985,
  },
  Object {
    "end": 2007,
    "name": "Mike Garnham",
    "share": 64,
    "start": 1986,
  },
  Object {
    "end": 2025,
    "id": 559,
    "member": 5004,
    "share": 32,
    "start": 2007,
  },
  Object {
    "end": 2025,
    "id": 1219,
    "member": 5004,
    "share": 32,
    "start": 2007,
  },
  Object {
    "current": true,
    "name": "John Smith",
    "share": 64,
    "start": 2025,
  },
]
`;

exports[`salesChanges no change 1`] = `Object {}`;

exports[`salesChanges none 1`] = `Object {}`;
